### PR TITLE
Add aten::_embedding_bag/_embedding_bag_forward_only

### PR DESCRIPTION
**Pull request by hjhee**
- _embedding_bag
- _embedding_bag_forward_only

### CODE REVIEW

**Comment by hjhee** [comment]
@fengyuan14

I'm not able to edit pull request to add new reviewer / add new assignees, please review this PR and check if granting such permissions are possible.

**Comment by hjhee** [comment]
failure cases are added to skipped lists:

- `AssertionError: Tensor-likes are not close!`: error are reproducible on current IPEX whl (ww21)
- `no out-of-bounds check on CUDA for perf.`: missing bound checks on CUDA / XPU kernel code (CPU only)
- `Could not run 'aten::_sparse_coo_tensor_with_dims_and_tensors'`: missing cases

**Comment by fengyuan14** [comment]
> failure cases are added to skipped lists:
> 
> * `AssertionError: Tensor-likes are not close!`: error are reproducible on current IPEX whl (ww21)
> * `no out-of-bounds check on CUDA for perf.`: missing bound checks on CUDA / XPU kernel code (CPU only)
> * `Could not run 'aten::_sparse_coo_tensor_with_dims_and_tensors'`: missing cases

Even the case fails on IPEX, we need to fix it. We need upstream bug free XPU implementation.

File: test/xpu/run_test_with_skip.py [hunk 1]
```diff
@@ -1030,31 +1038,37 @@ def launch_test(test_case, skip_list=None, exe_list=None):
...
-    "test_embedding_bag_device_xpu_int32_int64_bfloat16",
-    "test_embedding_bag_device_xpu_int32_int64_float16",
-    "test_embedding_bag_device_xpu_int32_int64_float32",
-    "test_embedding_bag_half_xpu_int32_int32",
-
-    # CPU fallback error: AssertionError: Tensor-likes are not close!
-    "test_EmbeddingBag_per_sample_weights_and_new_offsets_xpu_int32_int64_bfloat16",
-    "test_EmbeddingBag_per_sample_weights_and_new_offsets_xpu_int64_int32_bfloat16",
-    "test_EmbeddingBag_per_sample_weights_and_new_offsets_xpu_int64_int64_bfloat16",
+    # skipped by CUDA for performance
```
Comment (fengyuan14) [inline test/xpu/run_test_with_skip.py:1046]:
```suggestion
    # CUDA backend has no such functionality in the operator.
    # skipped by CUDA for performance
```

File: src/aten/sycl/EmbeddingBagKernel.h [hunk 2]
```diff
@@ -0,0 +1,591 @@
...
+template <
+    typename scalar_t,
+    typename accscalar_t,
+    typename index_t,
+    int mode,
+    int vec_size,
+    typename vec_t,
+    typename vec_acc_t,
+    typename vec_idx_t>
+struct EmbeddingBagKernelFunctor {
```
Comment (fengyuan14) [inline src/aten/sycl/EmbeddingBagKernel.h:27]: It should be the interface header of kernels. Please remove functor definition to EmbeddingBag.cpp

File: src/aten/sycl/EmbeddingBag.cpp [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+#undef VEC_EMBBAG_KERNEL
+}
+
+} // namespace
+
+std::tuple<Tensor, Tensor, Tensor, Tensor> _embedding_bag_kernel(
+    const Tensor& weight_t,
+    const Tensor& indices_t,
+    const Tensor& offsets_t,
+    const bool scale_grad_by_freq,
```
Comment (fengyuan14) [inline src/aten/sycl/EmbeddingBag.cpp:554]: Don't we support the semantics in kernel? Please review CUDA kernel implementation to get it supported.

File: src/aten/sycl/EmbeddingBag.cpp [hunk 4]
```diff
@@ -0,0 +1,682 @@
...
+  if (indices_t.dim() == 1) {
+    TORCH_CHECK(
+        offsets_t.dim() == 1,
+        "offsets has to be a 1D Tensor, but got Tensor of dimension ",
+        offsets_t.dim());
+  }
+  TORCH_CHECK(
+      weight_t.dim() == 2,
+      "weight has to be a 2D Tensor, but got Tensor of dimension ",
+      weight_t.dim());
```
Comment (fengyuan14) [inline src/aten/sycl/EmbeddingBag.cpp:573]: Remove codes to `src/aten/EmbeddingBag.cpp`. It should be operator level related.

File: src/aten/sycl/EmbeddingBag.cpp [hunk 5]
```diff
@@ -0,0 +1,682 @@
...
+      TORCH_CHECK(0, "Invalid EmbeddingBag mode (max, sum, mean) ...");
+  };
+#undef EXTEND_EMBBAG_TEMPLATE
+#else
+  int64_t weight_total_elem = weight.numel();
+  AT_DISPATCH_FLOATING_TYPES_AND2(
+      at::ScalarType::Half,
+      at::ScalarType::BFloat16,
+      weight.scalar_type(),
+      "embedding_bag_kernel",
```
Comment (fengyuan14) [inline src/aten/sycl/EmbeddingBag.cpp:648]:
```suggestion
      "embedding_bag_xpu",
```

File: src/aten/sycl/EmbeddingBag.cpp (hunk 3 above)
Comment (fengyuan14) [reply src/aten/sycl/EmbeddingBag.cpp:554]: Checked. Not used in forward.

**Comment by fengyuan14** [comment]
Case coverage,
image

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**